woocommerce/wc-admin
# Components: Add a search component based on Autocomplete

This is an alternate approach to the search in #295, which uses `FormTokenField`. I was having trouble getting `FormTokenField` to consistently return results, and the results it returns don't include the product ID, just the name of the product. I decided to try the approach suggested in https://github.com/woocommerce/wc-admin/issues/257#issuecomment-412275723, of using Gutenberg's Autocompleter.

The core (gb) Autocomplete component expects to be used with a contentEditable element, and expects multiple autocompleters which can be triggered with a prefix (`@` for users, `+` for p2s, etc). It also defaulted to replacing the search content, but you could type anything around the replaced tokens. On the other hand, the format of the autocompleter objects ([see gb's user as an example](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/autocompleters/user.js)) would be much easier to extend to other object types than the `FormTokenField` in #295, and we have much more control over the look of the responses.

So I've taken the base of `<Autocomplete />`, and re-implemented it as [`search/autocomplete.js`](https://github.com/woocommerce/wc-admin/blob/try/autocomplete-search/client/components/search/autocomplete.js). This is an "internal-only" component, used by the Search component, and it uses the selected autocompleter to fetch search results. Eventually, we could add orders & customers handlers, and it would be as simple as creating a new file in [`autocompleters/`](https://github.com/woocommerce/wc-admin/tree/try/autocomplete-search/client/components/search/autocompleters). We could also add a way to pass in an autocompleter object, so that 3rd party devs can easily add searches for their data too.

**Screenshots**

Searching a term with matches:

![search-with-results](https://user-images.githubusercontent.com/541093/44630757-e62c7e00-a930-11e8-9830-8c92dd601d78.png)

After selecting two items:

![search-with-tags](https://user-images.githubusercontent.com/541093/44630758-e62c7e00-a930-11e8-9cac-8fb7929975d3.png)

This can also be used in the advanced filter's product selection, but it'll need some work:

![needs-work](https://user-images.githubusercontent.com/541093/44630759-e62c7e00-a930-11e8-9b50-dbdb01b1c424.png)

This PR also adds a `<Tag />` component, which is used to display the selected result. Both Search & Tag have readmes:

- [Search README](https://github.com/woocommerce/wc-admin/tree/try/autocomplete-search/client/components/search)
- [Tag README](https://github.com/woocommerce/wc-admin/tree/try/autocomplete-search/client/components/tag)

**To test**

I've added a demo to the Analytics > Overview page, so you can test it out there.

- Search for a product, make sure selecting it creates a Tag
- Remove the Tag by clicking the X

This is basically a proof-of-concept, so if it looks good I can come back & add in the loading state, a no-results state, etc. Maybe an "inline" version, with search & tokens on the same line (needed for the Advanced filter).
